R/module_teal_data.R

```diff
@@ -250,3 +219,33 @@ srv_check_shiny_warnings <- function(id, data, modules) {
     })
   })
 }
+
+
+#' Fallback on failure
+#'
+#' Function returns the previous reactive if the current reactive is invalid (throws error or returns NULL).
+#' Application: In `teal` we try to prevent the error from being thrown and instead we replace failing
+#' transform module data output with data input from the previous module (or from previous `teal` reactive
+#' tree elements).
+#'
+#' @param this (`reactive`) Current reactive.
+#' @param that (`reactive`) Previous reactive.
+#' @param label (`character`) Label for identifying problematic `teal_data_module` transform in logging.
+#' @return `reactive` `teal_data`
+#' @keywords internal
+.fallback_on_failure <- function(this, that, label) {
+  assert_reactive(this)
+  assert_reactive(that)
+  checkmate::assert_string(label)
+
+  reactive({
+    res <- try(this(), silent = TRUE)
+    if (inherits(res, "teal_data")) {
+      logger::log_debug("{ label } evaluated successfully.")
+      res
+    } else {
+      logger::log_debug("{ label } failed, falling back to previous data.")
+      that()
+    }
+  })
+}
```

Comment:
I think that `.fallback_on_failure` could look differently. Previously it was used in a way that each stage has `.fallback_on_failure` `teal_data_module -> .fallback_on_failure -> teal_transform_module[1] -> .fallback_on_failure -> teal_transform_module[2] -> ...`
This meant that if `teal_transform_module[2]` failed it took data output from `teal_transform_module[1]`. If `teal_transform_module[2]` failed as well then `teal_data_module` output was rolled. We can talk more on a daily 👍 

```suggestion
.fallback_on_failure <- function(this, that, label) {
  assert_reactive(this)
  assert_reactive(that)
  checkmate::assert_string(label)
  out <- reactiveVal()
  observe({
    if (inherits(res, "teal_data")) {
      logger::log_debug("{ label } evaluated successfully.")
      shiny::isolate(out(this()))
    }
  })
  out
}
```

Reply:
Should 

```r
if (inherits(res, "teal_data") && !identical(this(), shiny::isolate(out()))) {
```

be 

```
if (inherits(that, "teal_data") && !identical(this(), shiny::isolate(out()))) {
```

so that `that` is used anywhere